### PR TITLE
Added API code for actions.

### PR DESCRIPTION
**Pull request by TimSimpsonR**
- Copied get_id_from_href from legacy nova.api.openstack.common to reddwarf.common.utils.
- Added stubs for resize_flavor, resize_volume, and restart into models.
- Copied with slight changes the API code for actions from legacy.
- Changed BadRequest to have the original message from legacy.
- Made a 'MissingKey' exception to easily use the new message that BadRequest had.

### CODE REVIEW

**Comment by ed-** [comment]
Sure, have a +1

File: reddwarf/instance/service.py [hunk 1]
```diff
@@ -80,6 +80,84 @@ def wrapper(*args, **kwargs):
 class InstanceController(BaseController):
     """Controller for instance functionality"""
 
+    def action(self, req, body, tenant_id, id):
+        LOG.info("req : '%s'\n\n" % req)
+        LOG.info("Comitting an ACTION again instance %s for tenant '%s'"
+                 % (id, tenant_id))
+        context = rd_context.ReddwarfContext(
```
Comment (hub-cap) [inline reddwarf/instance/service.py:87]: new context?

**Comment by hub-cap** [comment]
hey this is great and fast. and great. and fast. and great. but how come the models is no implemented?

**Comment by cp16net** [comment]
+1

File: reddwarf/instance/models.py [hunk 2]
```diff
@@ -216,6 +217,41 @@ def _validate_can_perform_action(self):
...
+        # TODO(tim.simpson): Validate old_size < new_size, or raise
+        #                    rd_exceptions.BadRequest.
+        # TODO(tim.simpson): resize volume.
+        raise RuntimeError("Not implemented (yet).")
+
+    def restart(self):
+        if instance_state in SERVER_INVALID_ACTION_STATUSES:
+            LOG.debug("Restart instance not allowed while instance is in %s "
+                      "status." % instance_state)
+            # If the state is building then we throw an exception back
```
Comment (rnirmal) [inline reddwarf/instance/models.py:238]: Hmm I like the debug message better than the one below. Can we change the exception message to say the same.

Comment (TimSimpsonR) [inline reddwarf/instance/models.py:238]: Good idea.

**Comment by TimSimpsonR** [comment]
I'll wait until the guest code is written a push again.

**Pull request closed without merging**